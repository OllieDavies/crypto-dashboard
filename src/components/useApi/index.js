import { useState, useCallback } from "react";
import { get } from "axios";

export default function useApi(path) {
  const [isLoading, setLoading] = useState(true);
  const [response, setResponse] = useState(undefined);
  const [error, setError] = useState(undefined);

  const fetch = useCallback(() => {
    setLoading(true);
    get(path)
      .then(setResponse)
      .catch(setError)
      .finally(() => setLoading(false));
  }, [path]);

  return {
    isLoading,
    response,
    error,
    fetch,
  };
}
