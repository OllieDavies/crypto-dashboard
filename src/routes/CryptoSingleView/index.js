import React, { useEffect } from "react";
import { useParams } from "react-router-dom";

import ChangeIndicator from "../../components/ChangeIndicator";
import MetaDisplay from "../../components/MetaDisplay";
import useApi from "../../components/useApi";

const CryptoSingle = () => {
  let { id } = useParams();

  const { isLoading, response, error, fetch } = useApi(
    `https://api.coinranking.com/v1/public/coin/${id}?base=GBP&timePeriod=7d`
  );

  // Fetch on mount and start a timer to poll
  useEffect(() => {
    fetch();
    const timer = setInterval(fetch, 3000);
    return () => clearInterval(timer);
  }, [fetch]);

  if (isLoading === true && response === undefined) return <h1>Loading...</h1>;
  if (error) return <h1>An error occured</h1>;

  const { coin } = response.data.data;

  return (
    <article>
      <h1>{coin.name}</h1>
      <MetaDisplay name="Price" value={`${coin.price} ${coin.symbol}`} />
      <MetaDisplay name="Market Cap" value={coin.marketCap} />
      <MetaDisplay
        name="Change"
        children={<ChangeIndicator change={coin.change} />}
      />
      <details open={true}>
        <summary>Coin Description</summary>
        <div dangerouslySetInnerHTML={{ __html: coin.description }} />
      </details>
    </article>
  );
};

export default CryptoSingle;
