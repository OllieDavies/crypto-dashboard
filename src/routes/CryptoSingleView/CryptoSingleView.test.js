import React from "react";
import { render } from "@testing-library/react";

import CryptoSingleView from ".";
import useApi from "../../components/useApi";

jest.mock("../../components/useApi");
jest.mock("react-router-dom", () => ({
  useParams: () => ({ id: 1 }),
}));

describe("<CryptoSingleView />", () => {
  it("matches snapshot when first mounted", () => {
    useApi.mockReturnValue({
      isLoading: true,
      fetch: jest.fn(),
    });
    const { container } = render(<CryptoSingleView />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot when error is returned", () => {
    useApi.mockReturnValue({
      isLoading: false,
      fetch: jest.fn(),
      error: new Error("Error"),
    });

    const { container } = render(<CryptoSingleView />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("matches snapshot when coin is returned", () => {
    useApi.mockReturnValue({
      isLoading: false,
      fetch: jest.fn(),
      response: {
        data: {
          data: {
            coin: {
              id: 1,
              name: "Bitcoin",
              symbol: "BTC",
              price: "7500",
              change: "0.3",
              description: "<p>Example description</p>",
            },
          },
        },
      },
    });

    const { container } = render(<CryptoSingleView />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
