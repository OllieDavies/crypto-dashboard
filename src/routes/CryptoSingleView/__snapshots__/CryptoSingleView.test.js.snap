// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CryptoSingleView /> matches snapshot when coin is returned 1`] = `
<article>
  <h1>
    Bitcoin
  </h1>
  <div>
    <p>
      <strong>
        Price
        :
      </strong>
       
      7500 BTC
    </p>
  </div>
  <div>
    <p>
      <strong>
        Market Cap
        :
      </strong>
       
    </p>
  </div>
  <div>
    <p>
      <strong>
        Change
        :
      </strong>
       
      <span
        class="text positive"
      >
        0.3
      </span>
    </p>
  </div>
  <details
    open=""
  >
    <summary>
      Coin Description
    </summary>
    <div>
      <p>
        Example description
      </p>
    </div>
  </details>
</article>
`;

exports[`<CryptoSingleView /> matches snapshot when error is returned 1`] = `
<h1>
  An error occured
</h1>
`;

exports[`<CryptoSingleView /> matches snapshot when first mounted 1`] = `
<h1>
  Loading...
</h1>
`;
